teps/0011-redirecting-step-inputs-and-outputs.md

```diff
@@ -0,0 +1,143 @@
+* Enhance the `entrypoint` binary to redirect stdin from and stdout/stderr to specified files.
+* Provide examples to use step redirection for the use cases mentioned in [Motivation](#motivation).
+
+## Proposal
+
+The following new fields will be added to the `Step` struct:
+
+```go
+type Step struct {
+    ...
+    // Path to redirect stdin from on container's local filesystem.
+    // +optional
+    Stdin string `json:"stdin,omitempty"`
+    // Path to redirect stdout to on container's local filesystem.
+    // +optional
+    Stdout string `json:"stdout,omitempty"`
+    // Path to redirect stderr to on container's local filesystem.
+    // +optional
+    Stderr string `json:"stderr,omitempty"`
+}
+```
+
+Once `Stdout` or `Stderr` is specified, the corresponding output stream will be duplicated to both the given file and the standard output stream of the container, so users can still view the output through the Pod log API. If both `Stdout` and `Stderr` are set to the same value, the two streams will be written to two file descriptors referring to the same file description, so outputs from both stream will be interleaved in the same file, but there will be no ordering guarantee on the data. The `Stdin` and `Stdout` (or `Stderr`) fields of a step cannot be set to the same value, or the specified file will be truncated and result in an empty stdin.  If multiple steps' `Stdout` fields are set to the same value, the file content will be overwritten by the last outputting step.
+
+Variable substitution will be applied to the new fields, so one could specify `$(results.<name>.path)` to the `Stdout` field to extract the stdout of a step into a Task result. No new variable substitution for accessing the values of `Stdout` and `Stderr` fields will be provided so variable substitution can remain single-pass.
+
+#### Example Usage
+
+Redirecting stdout of `boskosctl` to `jq` and publish the resulting `project-id` as a Task result:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: boskos-acquire
+spec:
+  results:
+  - name: project-id
+  steps:
+  - name: boskosctl
+    image: gcr.io/k8s-staging-boskos/boskosctl
+    args:
+    - acquire
+    - --server-url=http://boskos.test-pods.svc.cluster.local
+    - --owner-name=christie-test-boskos
+    - --type=gke-project
+    - --state=free
+    - --target-state=busy
+    stdout: /data/boskosctl-stdout
+    volumeMounts:
+    - name: data
+      mountPath: /data
+  - name: parse-project-id
+    image: imega/jq
+    args:
+    - -r
+    - '.name'
+    stdin: /data/boskosctl-stdout
+    stdout: $(results.project-id.path)
+    volumeMounts:
+    - name: data
+      mountPath: /data
+  volumes:
+  - name: data
+```
+
+### Risks and Mitigations
+
+* Users might mistakenly specify paths not shared among steps for redirection. Potentially Tekton could implement a validation to warn users about such misuses.
```

Comment:
Indeed, that might be a problem. We may want to restrict it to `workspaces` but it's gonna be tricky to validate

Reply:
I think this is an argument for just always writing stdout/stderr to some Tekton-owned path and documenting where that is. If a later step wants to read it, they can.

Reply:
Yeah this is why I explicitly call this out in this section so we can discuss if "conventional paths" as you suggest is a better way.

Reply:
On the other hand, if we make it explicit that this is a path, most users would naturally be aware that if they want to use the same path in different containers, they have to make a shared mount point.